src/storage/kv/transaction.rs

```diff
@@ -542,7 +659,56 @@ impl Transaction {
         self.savepoints = 0;
         self.write_seqno = 0;
     }
+}
 
+struct WriteSetRangeIterator<'a> {
+    // Use a sorted vec of references instead of direct hash map iteration
+    entries: Vec<(&'a Bytes, &'a WriteSetEntry)>,
+    current: usize,
+}
+
+impl<'a> WriteSetRangeIterator<'a> {
+    fn new<R>(write_set: &'a HashMap<Bytes, Vec<WriteSetEntry>>, range: R) -> Self
+    where
+        R: RangeBounds<&'a [u8]>,
+    {
+        // Collect and sort all valid entries up front
+        let mut entries: Vec<_> = write_set
+            .iter()
+            .filter_map(|(key, entries)| {
+                if !Transaction::is_key_in_range(key, &range) {
+                    return None;
+                }
+                entries.last().map(|entry| (key, entry))
+            })
+            .collect();
+
+        // Sort by key
+        entries.sort_by(|a, b| a.0.cmp(b.0));
```

Comment:
It think it might actually be better to change the Write Set from `HashMap` to `BTreeMap`. That way you won't have to collect the latest values upfront in a vector and sort them, since `BTreeMap`'s iterator is ordered. Maybe this will be enough to keep the scan perf in par with the current one.